aten/src/ATen/templates/Functions.h

```diff
@@ -20,14 +21,22 @@ using native::tensor;
 
 ${function_declarations}
 
-static inline Type & infer_type(const Tensor & t) {
+namespace detail {
+
+static inline TypeExtendedInterface & infer_type(const Tensor & t) {
   AT_CHECK(t.defined(), "undefined Tensor");
-  return t.type();
+  return static_cast<TypeExtendedInterface&>(t.type());
```

Comment:
see below, I think we should have a non-core interface that returns TypeExtendedInterface and we should call that instead of type() and casting.